### PR TITLE
Decouple fitting with differentiable input.

### PR DESCRIPTION
**Pull request by bejaeger**
The API allowed to fit w/ differentiable inputs, i.e. also accepted `torch.Tensor`'s  in the standard `fit()` call. This bloated the code with several `if/else` statements. The following decouples the paths which allows to be more explicit with type hints.

---

## Public API Changes

-   [ ] No Public API changes
-   [X] Yes, Public API changes (Details below)

---

## How Has This Been Tested?

---

## Checklist

-   [X] The changes have been tested locally.
-   [X] Documentation has been updated (if the public API or usage changes).
-   [X] A changelog entry has been added (see `changelog/README.md`), or "no changelog needed" label requested.
-   [X] The code follows the project's style guidelines.
-   [X] I have considered the impact of these changes on the public API.

---

### CODE REVIEW

File: src/tabpfn/classifier.py [hunk 1]
```diff
@@ -786,61 +831,46 @@ def fit(
...
-        )
-        if is_differentiable_input_and_already_fitted:
+        is_first_fit_call = not hasattr(self, "models_")
+        if is_first_fit_call:
+            byte_size, rng = self._initialize_model_variables()
+            ensemble_configs, X, y = self._initialize_for_differentiable_input(
+                X=X, y=y, rng=rng
+            )
+            self.ensemble_configs_ = ensemble_configs  # Store for prompt tuning reuse
+            remove_non_differentiable_preprocessing_from_models(models=self.models_)
```
[the remark this answers is not in the document]
Comment (bejaeger) [reply src/tabpfn/classifier.py:849]: Same behavior as before and this should be fine.

File: src/tabpfn/classifier.py [hunk 2]
```diff
@@ -786,61 +831,46 @@ def fit(
         Returns:
             self
         """
```
[the remark this answers is not in the document]
Comment (bejaeger) [reply src/tabpfn/classifier.py:833]: Let's take type hints more seriously and avoid this code bloat

File: changelog/752.added.md [hunk 3]
```diff
@@ -0,0 +1,2 @@
+- The estimators' `fit()`/`predict()` functions now only allow the types `pd.DataFrame` and `np.ndarray`
+- The classifier gets a dedicated method for fitting with differentiable input called `fit_with_differentiable_input()`
```
[the remark this answers is not in the document]
Comment (bejaeger) [reply changelog/752.added.md]: Since it's experimental code, I think we don't need to add specific predict_with_differentiable_input functions.

File: tests/test_consistency.py [hunk 4]
```diff
@@ -207,7 +207,10 @@ def test__fit_predict__predictions_match_reference(
 def _predict(test_case: _ConsistencyCase) -> np.ndarray:
     X_train, y_train, X_test = test_case.data()
     model = test_case.model()
-    model.fit(X_train, y_train)
+    if test_case.model().differentiable_input:
```
Comment (psinger-prior) [inline tests/test_consistency.py]: Could re-use `model` here

**Comment by bejaeger** [comment]
Thanks for reviewing and trying it out. The issue should be fixed!

**Review by psinger-prior — APPROVED** [review]
LGTM!

Only thing I see a lot are these warnings:
```
UserWarning: gemm_and_bias error: CUBLAS_STATUS_NOT_INITIALIZED when calling cublasLtMatmul with transpose_mat1 1 transpose_mat2 0 m 10 n 88 k 384 mat1_ld 384 mat2_ld 384 result_ld 10 abType 0 cType 0 computeType 68 scaleType 0. Will attempt to recover by calling unfused cublas path. (Triggered internally at /pytorch/aten/src/ATen/cuda/CUDABlas.cpp:1765.)
  return F.linear(input, self.weight, self.bias)
```

Wonder if that can be addressed easily.

**Comment by bejaeger** [comment]
Hmm...I'm not seeing this kind of warning. Tried a Mac & linux system.
Merging for now, thanks for the review!
